New gateway: HSBC Direct Debit API for HK

bruno opened the pull request:
This Pull Request introduces a new payment gateway (HSBC HK) that *only* supports an API for direct debit authorisations. This is only for Hong Kong.

We've also introduced a `Dockerfile` and associated tooling to enable local development using docker.

henare committed:
Add a Dockerfile and instructions for local development using Docker

Refs waysact/evergiving#6651

henare committed:
Generate scaffold of new HSBC gateway

Using:

`docker run -it --rm -v "$PWD":/app waysact/active_merchant ./script/generate gateway hsbc`

Refs waysact/evergiving#6651

henare committed:
Set test_url for HSBC gateway

Refs waysact/evergiving#6651

henare committed:
Set HSBC gateway metadata

Refs waysact/evergiving#6651

henare committed:
Set required credentials for HSBC gateway

Refs waysact/evergiving#6651

henare committed:
Get test_invalid_login working for HSBC gateway

Refs waysact/evergiving#6651

henare committed:
WIP

bruno committed:
WIP send and receive encrypted payloads to HSBC

This work-in-progress adds a number of changes to the HSBC direct debit authorisation implementation:

- an `ActiveMerchant::Crypto` class that uses `gpgme` to encrypt and sign a payload prior to sending to HSBC, and decrypt and verify the ciphertext we get back from the response;
- small changes to the `add_direct_debit_authorisation_data` method that builds the payload;
- ensure that we always send the `application/json` content type header;

I've also redacted the remote test data, as it contains details that should probably not be in a public repo. The fixures changes that contain the development GPG public and private keys are also not committed.

The base docker container we use contains gpg and the gpgme required binaries to build the gem 🎉

Refs waysact/evergiving#6651

bruno committed:
Fix test assertion

Refs waysact/evergiving#6651

bruno committed:
Add OTP confirmation method and small refactor

Add the ability to confirm the authorisation of the direct debit by calling the `authorisations/otp-confirmation` API endpoint with the data supplied via the SMS and API. I don't believe we have a test OTP password/mandate identification that we can use without getting a proper 2FA via SMS and entering it interactively, so I had to implement it with the `STDIN.gets` method in the remote test.

Did a bit of refactor to DRY the code that calls the ActiveMerchant::Crypto module (and fixed a comment in that one).

These changes have enabled me to generate a successful direct debit authorisation and confirm it, thus getting an SMS from the pre-prod system that states that HSBC received the direct debit authorisation setup request.

Refs waysact/evergiving#6651

bruno committed:
Fix typo

Refs waysact/evergiving#6651

henare committed:
Add ability to regenerate OTP code for HSBC gateway

Refs waysact/evergiving#6651

henare committed:
Scrub client ID and secret from transcripts for HSBC gateway

Refs waysact/evergiving#6651

bruno committed:
Remove hardcoded recipients and signers

These two values are not necessary as GPGME will use extract the values from the keys supplied. This also allows us to use development keys for the unit tests.

Refs waysact/evergiving#6651

bruno committed:
Add development gpg key pair

These keys were generated exclusively for this ActiveMerchant implementation, to support the unit tests.

**these keys should never be used in any environment outside of running the test suite**.

Refs waysact/evergiving#6651

bruno committed:
Return authorisation from HSBC response

In the case of the Direct Debit API, the authorisation equivalent is the MandateIdentification, so we return this in the `authorization_from` method.

Refs waysact/evergiving#6651

bruno committed:
Add first unit test for HSBC authorization

Adds support for our first unit test in the HSBC implementation.

Refs waysact/evergiving#6651

bruno committed:
Mark two tests as pending implementation

In order to avoid failures instead, mark them as pending.

Refs waysact/evergiving#6651

henare committed:
Remove items in the response that we're not generating

Refs waysact/evergiving#6651

henare committed:
Add test for failed authorisation and modify error handling

Refs waysact/evergiving#6651

henare committed:
Test transcript scrubbing on HSBC gateway

Refs waysact/evergiving#6651

henare committed:
Revert "Remove hardcoded recipients and signers"

This reverts commit cef732fe9695899a0f8357d2597eb74e64722176.

henare committed:
Add private_key to required parameters

Refs waysact/evergiving#6651

henare committed:
Don't hardcode recipient and signer for encrypt_and_sign

Refs waysact/evergiving#6651

bruno committed:
private_key.imports contains 2 elements

The private key also contains the public key (h/t @henare for pointing me to https://superuser.com/questions/434614/do-pgp-private-key-blocks-contain-the-public-key-as-well), so let's ensure that `imports.size` has 2 elements.

Refs waysact/evergiving#6651

henare committed:
Use correct variable name

Refs waysact/evergiving#6651

henare committed:
Add comment showing the unencrypted payload

This makes it a lot easier to understand what's going on.

Refs waysact/evergiving#6651

henare committed:
Add unit test for successful OTP confirmation with HSBC gateway

Refs waysact/evergiving#6651

henare committed:
Also check the OTP ID in HSBC authorisation success unit test

Refs waysact/evergiving#6651

henare committed:
Change assertion to something more specific

Refs waysact/evergiving#6651

henare committed:
Add unit test for successful OTP regeneration on HSBC gateway

Refs waysact/evergiving#6651

henare committed:
Add unit test failure when wrong OTP code is provided to HSBC gateway

Refs waysact/evergiving#6651

henare committed:
Fix indentation only

Refs waysact/evergiving#6651

henare committed:
Add unit test for failed OTP regeneration on HSBC gateway

Refs waysact/evergiving#6651

henare committed:
Fix test assertions

We were asserting the wrong variable. That's one way to get the tests to
pass! Oops.

Refs waysact/evergiving#6651

henare committed:
Call raise properly by passing the class and message as parameters

Refs waysact/evergiving#6651

jscheid reviewed (commented):
Nice work Henare. I worry a little about the crypto performance, considering that it appears to be used per-request. Do you have an idea of its overhead?

bruno reviewed (commented):
Great job on taking this @henare ... It does look good to me.

henare commented:
> Nice work Henare. I worry a little about the crypto performance, considering that it appears to be used per-request. Do you have an idea of its overhead?

Good point @jscheid - we haven't tested crypto performance at all so far. I also note this implementation writes to disk as part of the crypto operations.

I just did some crude tests and on my machine and it seems to take between 60 and 90ms to encrypt a payload. Decrypting was sub-millisecond. Given that, is it something we should look into more?

fbrusatti commented:
Outstanding work here guys 👏 . 

If performance of encryption/decryption is an issue, a very small tweak could be decrypt only once in `commit` method and avoid decrypt twice (because `message_from` where you decrypt is called twice)

⚠️ remember before push this to change the `live_url` (I did that mistake before)

fbrusatti commented on lib/active_merchant/billing/gateways/hsbc.rb:
```diff
@@ -0,0 +1,210 @@
+        super
+      end
+
+      def authorize(money, payment, options={})
```
is `payment` needed here? Is it going to be used in a future? I think we don't use it in anyplace now.

henare replied:
Nice catch! Fixed in https://github.com/waysact/active_merchant/pull/157/commits/09e9d5a9f704fe403e23459e41aa592781aa9046.

jscheid commented:
I reckon 60-90ms overhead per payment isn't great, but acceptable. Maybe something that could be improved down the road. If it's multiples of 60-90ms per payment (not sure how many requests there are and how many of them need encryption) then maybe it's worth looking into it sooner than later.

henare committed:
Remove unused payment parameter

H/T to @fbrusatti's review :star:

Refs waysact/evergiving#6651

henare committed:
Fix test_invalid_login remote test for HSBC

This wasn't working because we now check for valid crypto keys.

When fixing this I discovered the error message wasn't being parsed
properly any more either. I've refined the parsing of the error and
improved the test by refining and expanding the assertions.

Refs waysact/evergiving#6651

henare commented:
> If performance of encryption/decryption is an issue, a very small tweak could be decrypt only once in `commit` method and avoid decrypt twice (because `message_from` where you decrypt is called twice)

@fbrusatti thanks for the review! That's a good catch that the decryption happens multiple times. In my crude performance tests decryption was so quick the overhead didn't matter so I don't think I'll worry about tweaking it now. Let me know if you disagree and I can take a look at it.

henare commented:
> I reckon 60-90ms overhead per payment isn't great, but acceptable. Maybe something that could be improved down the road. If it's multiples of 60-90ms per payment (not sure how many requests there are and how many of them need encryption) then maybe it's worth looking into it sooner than later.

Thanks @jscheid, it sounds like we can look at improving this down the road.

There are two requests per normal payment but these requests are separated by the OTP code being sent to the donor and then entered so the compounded overhead doesn't impact the experience.
